binder/src/test/java/io/grpc/binder/internal/ActiveTransportTrackerTest.java

```diff
@@ -0,0 +1,138 @@
+
+import static org.junit.Assert.assertThrows;
+
+import android.os.Binder;
+import android.os.IBinder;
+import com.google.common.collect.ImmutableList;
+import io.grpc.Attributes;
+import io.grpc.Metadata;
+import io.grpc.internal.FixedObjectPool;
+import io.grpc.internal.ObjectPool;
+import io.grpc.internal.ServerListener;
+import io.grpc.internal.ServerStream;
+import io.grpc.internal.ServerTransport;
+import io.grpc.internal.ServerTransportListener;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.robolectric.RobolectricTestRunner;
+
+@RunWith(RobolectricTestRunner.class)
+public final class ActiveTransportTrackerTest {
+
+  private static final OneWayBinderProxy.Decorator NO_OP_BINDER_DECORATOR = input -> input;
+
+  private final ObjectPool<ScheduledExecutorService> executorServicePool =
+      new FixedObjectPool<>(Executors.newSingleThreadScheduledExecutor());
+  private final IBinder callbackBinder = new Binder();
+
+  private boolean serverShutdownNotified = false;
+  private final Runnable testShutdownListener = () -> serverShutdownNotified = true;
+  private ActiveTransportTracker tracker;
+
+  @Before
+  public void setUp() {
+    tracker = new ActiveTransportTracker(new NoOpServerListener(), testShutdownListener);
+  }
+
+  @Test
+  public void testServerShutdown_onlyNotifiesAfterAllTransportAreTerminated() {
+    ServerTransportListener wrapperListener1 = registerNewTransport();
+    ServerTransportListener wrapperListener2 = registerNewTransport();
+
+    tracker.serverShutdown();
+    // 2 active transports, notification scheduled
+    assertThat(serverShutdownNotified).isFalse();
+
+    wrapperListener1.transportTerminated();
+    // 1 active transport remaining, notification still pending
+    assertThat(serverShutdownNotified).isFalse();
+
+    wrapperListener2.transportTerminated();
+    // No more active transports, shutdown notified
+    assertThat(serverShutdownNotified).isTrue();
+  }
+
+  @Test
+  public void testServerShutdown_noActiveTransport_notifiesTerminationImmediately() {
+    checkState(!serverShutdownNotified);
+
+    tracker.serverShutdown();
+
+    assertThat(serverShutdownNotified).isTrue();
+  }
+
+  @Test
+  public void testLastTransportTerminated_serverNotShutdownYet_doesNotNotify() {
+    ServerTransportListener wrapperListener = registerNewTransport();
+    assertThat(serverShutdownNotified).isFalse();
+
+    wrapperListener.transportTerminated();
+
+    assertThat(serverShutdownNotified).isFalse();
+  }
+
+  @Test
+  public void testTransportCreation_afterServerShutdown_throws() {
+    tracker.serverShutdown();
+
+    assertThrows(IllegalStateException.class, this::registerNewTransport);
+  }
+
+  private ServerTransportListener registerNewTransport() {
+    BinderTransport.BinderServerTransport transport =
```

Comment:
Could this be a fake or mock io.grpc.internal.ServerTransport instead? If so, you wouldn't need to come up with all the deps needed to create a real BinderServerTransport.

Reply:
There's some internal guidance for preferring real objects over test doubles. I think the mocks are fine for the cases where we need to verify interaction and we don't want to track that ourselves, but this objects seemed relatively test-friendly.

Kept it for now. Let me know if you feel strongly about it.

Reply:
Agreed that in general we prefer to test state rather than interactions (go/tott/281). But in this case, the real BinderServerTransport doesn't expose the state you need in order to test ActiveTransportTracker's function as a `ServerTransportListener` decorator (See go/choose-test-double#interaction-verification-appropriate case reason 2 and 3) In particular, you need to verify that, in addition to the SUT's tracking function, every ServerTransportListener method call results in exactly one method call on the same thread with the same arguments to the delegate ServerTransportListener before returning. BinderServerTransport doesn't record or expose any of the state you'd need to verify this (why would it). OTOH, mocks are perfect for testing decorators like this.

LMK if you buy this.